modules/dev-utils/src/main/java/org/apache/ignite/development/utils/IgniteIndexReader.java

```diff
@@ -0,0 +1,2130 @@
+    /**
+     * Read index file.
+     */
+    public void readIdx() {
+        long partPageStoresNum = Arrays.stream(partStores)
+            .filter(Objects::nonNull)
+            .count();
+
+        print("Partitions files num: " + partPageStoresNum);
+
+        Map<Class, AtomicLong> pageClasses = new HashMap<>();
+
+        long pagesNum = idxStore == null ? 0 : (idxFile.length() - idxStore.headerSize()) / pageSize;
+
+        print("Going to check " + pagesNum + " pages.");
+
+        Map<Class, Set<Long>> pageIoIds = new HashMap<>();
+
+        AtomicReference<Map<String, TreeTraversalInfo>> treeInfo = new AtomicReference<>();
+
+        AtomicReference<Map<String, TreeTraversalInfo>> horizontalScans = new AtomicReference<>();
+
+        AtomicReference<PageListsInfo> pageListsInfo = new AtomicReference<>();
+
+        List<Throwable> errors;
+
+        ProgressPrinter progressPrinter = new ProgressPrinter(System.out, "Reading pages sequentially", pagesNum);
+
+        try {
+            errors = scanFileStore(INDEX_PARTITION, FLAG_IDX, idxStore, (pageId, addr, io) -> {
+                progressPrinter.printProgress();
+
+                pageClasses.computeIfAbsent(io.getClass(), k -> new AtomicLong(0)).incrementAndGet();
+
+                if (io instanceof PageMetaIO) {
+                    PageMetaIO pageMetaIO = (PageMetaIO)io;
+
+                    long metaTreeRootId = normalizePageId(pageMetaIO.getTreeRoot(addr));
+
+                    treeInfo.set(traverseAllTrees("Index trees traversal", metaTreeRootId, CountOnlyStorage::new, this::traverseTree));
+
+                    treeInfo.get().forEach((name, info) -> {
+                        info.innerPageIds.forEach(id -> {
+                            Class cls = name.equals(META_TREE_NAME)
+                                ? IndexStorageImpl.MetaStoreInnerIO.class
+                                : H2ExtrasInnerIO.class;
+
+                            pageIoIds.computeIfAbsent(cls, k -> new HashSet<>()).add(id);
+                        });
+
+                        pageIoIds.computeIfAbsent(BPlusMetaIO.class, k -> new HashSet<>()).add(info.rootPageId);
+                    });
+
+                    Supplier<ItemStorage> itemStorageFactory = checkParts ? LinkStorage::new : CountOnlyStorage::new;
+
+                    horizontalScans.set(traverseAllTrees("Scan index trees horizontally", metaTreeRootId, itemStorageFactory, this::horizontalTreeScan));
+                }
+                else if (io instanceof PagesListMetaIO)
+                    pageListsInfo.set(getPageListsMetaInfo(pageId));
+                else {
+                    if (indexes == null) {
+                        ofNullable(pageIoIds.get(io.getClass())).ifPresent((pageIds) -> {
+                            if (!pageIds.contains(pageId)) {
+                                boolean foundInList =
+                                    (pageListsInfo.get() != null && pageListsInfo.get().allPages.contains(pageId));
```

Comment:
Not need to mention as orphan if page found in reuse list.
I have got the warning when destroyed sql cache in test.